### PR TITLE
Fix addGitLabMRComment

### PR DESCRIPTION
**Pull request by Alceatraz**
Recommit of #1315, Old one is too old to merge.

### CODE REVIEW

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Most maintainers (like me) are unlikely to merge a pull request that adds functionality and includes no test of the added functionality.  When new functionality is added without a test, that places responsibility on the maintainer to verify the code is still functioning before each release.  That quickly becomes more than a maintainer can do.

Can you add one or more tests that cover the newly added code?

**Comment by Alceatraz** [comment]
> Most maintainers (like me) are unlikely to merge a pull request that adds functionality and includes no test of the added functionality. When new functionality is added without a test, that places responsibility on the maintainer to verify the code is still functioning before each release. That quickly becomes more than a maintainer can do.
> 
> Can you add one or more tests that cover the newly added code?

I can understand but that become a problem for me, I do test and infact using this fork in production. But I have no idea how to write pipeline and create gitlab repo in an junit case.
I do willing provide *things* if you need, Is there any other way to prove this patch working ?

**Comment by krisstern** [comment]
Hi @Alceatraz Could you add the test(s) needed for this PR if possible? Otherwise this cannot be reviewed.

**Comment by Alceatraz** [comment]
> Hi @Alceatraz Could you add the test(s) needed for this PR if possible? Otherwise this cannot be reviewed.

I can't.

This feature need at least two nested jenkins pipeline and gitlab setting, Which can't set up by junit process. In my real usage, Jenkins nest depth is 4, Each of them need Add "compile success" comment into gitlab pipeline. ( Because the PR for "approve PR" is pending for 3 years now )

I can provied screenshot. But that seem like unacceptable when last time Mark asking me for tests. If so, Just close this PR. Sorry I can't help.

**Comment by krisstern** [comment]
@Alceatraz No worries, now I see your pain point for adding tests. Please do provide the screenshot for me so I can check before considering merging this PR. 

Thanks very much for your contribution!

**Comment by krisstern** [comment]
@MarkEWaite Is it okay for me to make an exception for adding test if after reviewing further I find the PR acceptable?

**Comment by MarkEWaite** [comment]
> @MarkEWaite Is it okay for me to make an exception for adding test if after reviewing further I find the PR acceptable?

Yes.  If it has been tested interactively and you're confident in the interactive testing, then merge without tests is a reasonable exception.  We want to be careful to not use that exception too frequently or we'll get into the situation we have with the git plugin that it needs interactive testing each time it is released.

**Comment by Alceatraz** [comment]
Ok, So sorry for the delay. Because the gitlab seems like brokened in recent version. Jenkins integration can't enable even with brand new install. Still fixing this issue. May be someone encounter this behavier before?

**Comment by Alceatraz** [comment]
Sorry for the long time delay.

I make a non-cut screen recording video to prove the patch is valid https://youtu.be/KtTkYgeihwY

And here is the key step

```groovy
pipeline {
    agent any
    stages {
        stage('test') {
            steps {
                addGitLabMRComment(comment: 'Parent Run')
                build('slave')
                addGitLabMRComment(comment: 'Parent Done')
            }
        }
    }
}
```

```groovy
pipeline {
    agent any
    stages {
        stage('test') {
            steps {
                addGitLabMRComment(comment: 'Slave Working')
            }
        }
    }
}
```


Using origin plugin:




Replace with my patched plugin:

**Comment by Alceatraz** [comment]
BTW: I will delete the youtube video when this PR merged.

**Comment by krisstern** [comment]
Thanks @Alceatraz for the follow-up! Let me test it and review this pull request soon

**Comment by krisstern** [comment]
> > @MarkEWaite Is it okay for me to make an exception for adding test if after reviewing further I find the PR acceptable?
> 
> Yes. If it has been tested interactively and you're confident in the interactive testing, then merge without tests is a reasonable exception. We want to be careful to not use that exception too frequently or we'll get into the situation we have with the git plugin that it needs interactive testing each time it is released.

Thanks @MarkEWaite

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
```

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
```

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
```

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
```

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
    private static List<GitLabWebHookCause> retrieveCauseRecursive(List<Cause> causes) {
```

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
                LOGGER.log(Level.WARNING, "Add MR comment failure, " +
                  "Cannot retrieve GitLab MR context: Cannot find GitLabWebHookCause");
```

**Review by krisstern — CHANGES_REQUESTED** [review]
Some editing requested, otherwise looks good!

**Comment by krisstern** [comment]
Will merge once CI checks have all been passed.

File: src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/workflow/AddGitLabMergeRequestCommentStep.java]:
```suggestion
     * Notice: Only retrieve the first found GitLabWebHookCause instance.
```

**Review by krisstern — APPROVED** [review]
Thanks @Alceatraz

**Comment by krisstern** [comment]
@MarkEWaite Is it okay to approve this pull request? I have checked that nothing seems out of place and it does what it is intended to do. But we will need your approval since you have requested changes to be made.

**Pull request merged**
